part_two.js

```diff
@@ -1 +1,164 @@
+
+function mapReport(store){
+  var result, report, reports, reportId, documents, pages, id;
+
+  reports = store.report;
+  documents = store.document;
+  pages = store.page;
+  result = {};
+
+  // this is assuming that there are checks upstream to ensure clean store val (ie, no dulipcates in this map)
+  for (id in pages){
+    report = reports[documents[pages[id]?.document_id]?.report_id]
+    if (report != undefined){
+      reportId = report.id
+      if (result[reportId]) {
+          ++result[reportId].number_of_pages
+      } else {
+          result[reportId]= {'report_title': report.title, 'report_id': reportId, 'number_of_pages': 1}
+      }
+    }
+  }
+  return result;
+}
+
+/* example function call:
+  mapReport(store)
+ --------------------- */
+
+/* Prompt 2 */
+  
+function mapStore (store){
+  var reports, documents, pages, parentReports;
+
+  reports = store.report;
+  documents = store.document;
+  pages = store.page;
+
+  parentReports = mapObjectsToParentReport( mapObjectsToParentReport(pages, documents, 'document_id'), reports, 'report_id');
+  return (Object.values(parentReports));
+}
+
+function mapObjectsToParentReport (childObject, parentObject, firstParentKeyType){
+  var childId, parent, parentId;
+
+  for (childId in childObject){
+    parentId = childObject[childId][firstParentKeyType]
+    parent = parentObject[parentId]
+    
+    if (parent != undefined){
+      if (parent.children == undefined ) {parent.children = []}
+      if (parent.report_id ){childObject[childId]['report_id'] = parent.report_id}
+      parent.children.push(childObject[childId])
+    }
+  }
+  return parentObject;
+}​
+
+function traverseReportMapBreadthFirst (parentQueue, searchString, searchIn ){
+  var next, search_result, report_ids_list;
+  report_ids_list =[];
+
+  while (parentQueue.length) {
+    next = parentQueue.pop()
+    
+    if (next){
+      search_result = searchObjectForString(next, searchString, searchIn)
+      if (search_result !== undefined){report_ids_list.push(search_result)}
+      else if (next.children !== undefined) {parentQueue = (next.children).concat(parentQueue)}
+    }
+  }
+  return report_ids_list;
+}
+
+function searchObjectForString (nextObject, searchString, searchIn){
+  var property, index;
+
+  for (property of searchIn){
+    index = nextObject[property]?.toLowerCase().indexOf(searchString)
+    if (index !== undefined && index !== -1){
+      return nextObject.report_id || nextObject.id;
+    } 
+  }
+}
+
+function searchString (store, searchString, searchIn){
+  var reportList, mappedStore;
+  reportList = [];
+  if (searchString) {
+    mappedStore = mapStore(store);
+    reportList.push(traverseReportMapBreadthFirst(mappedStore, searchString.toLowerCase(), searchIn));
+    return reportList;
+  }
+}
+
+/* example function call:
+
+var fieldsToSearch = ['title', 'name', 'body', 'footnote']
+var query = 'quart'
+searchString(store, query, fieldsToSearch)
+ --------------------- */
```

Comment:
decided to add a field searchIn to specify which fields to consider/ exclude instead of hardcoding which fields to ignore (such as filetype). This also gives us a way to customize/filter if we wanted to eventually give a user that ability.